arch/riscv/core/riscv_smp.c

```diff
@@ -0,0 +1,219 @@
+/*
+ * Copyright (c) 2021 Microchip Inc.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/**
+ * @file
+ * @brief codes required for Risc-V multicore and Zephyr smp support
+ *
+ */
+#include <device.h>
+#include <kernel.h>
+#include <kernel_structs.h>
+#include <ksched.h>
+#include <soc.h>
+#include <init.h>
+#include <rv_smp_defs.h>
+
+volatile struct {
+	arch_cpustart_t fn;
+	void *arg;
+} riscv_cpu_init[CONFIG_MP_NUM_CPUS];
+
+/*
+ * Collection of flags to control wake up of harts. This is trickier than
+ * expected due to the fact that the wfi can be triggered when in the
+ * debugger so we have to stage things carefully to ensure we only wake
+ * up at the correct time.
+ *
+ * initial implementation which assumes any monitor hart is hart id 0 and
+ * SMP harts have contiguous hart IDs. CONFIG_SMP_BASE_CPU will have minimum
+ * value of 1 for systems with monitor hart and zero otherwise.
+ * 
+ */
+
+#define WAKE_FLAG_COUNT (CONFIG_MP_TOTAL_NUM_CPUS)
+
+/* we will index directly off of mhartid so need to be careful... */
+volatile __noinit uint64_t hart_wake_flags[WAKE_FLAG_COUNT];
+
+volatile char *riscv_cpu_sp;
+/*
+ * _curr_cpu is used to record the struct of _cpu_t of each cpu.
+ * for efficient usage in assembly
+ */
+volatile _cpu_t *_curr_cpu[CONFIG_MP_NUM_CPUS];
+
+/* Called from Zephyr initialization */
+void arch_start_cpu(int cpu_num, k_thread_stack_t *stack, int sz,
+		    arch_cpustart_t fn, void *arg)
+{
+	int hart_num = cpu_num + CONFIG_SMP_BASE_CPU;
+
+	/* Used to avoid empty loops which can cause debugger issues
+	 * and also for retry count on interrupt to keep sending every now and again...
+	 */
+	volatile int counter = 0; 
+
+	_curr_cpu[cpu_num] = &(_kernel.cpus[cpu_num]);
+	riscv_cpu_init[cpu_num].fn = fn;
+	riscv_cpu_init[cpu_num].arg = arg;
+
+	/* set the initial sp of target sp through riscv_cpu_sp
+	 * Controlled sequencing of start up will ensure
+	 * only one secondary cpu can read it per time 
+	 */
+	riscv_cpu_sp = Z_THREAD_STACK_BUFFER(stack) + sz;
+
+	/* wait slave cpu to start */
+	while (hart_wake_flags[hart_num] != RV_WAKE_WAIT) {
```

Comment:
what if the secondary core fails to start (for any reason)? Shall we handle such case?